Добавил статью for...in (#1578)
doka-guide/content: js/for-in/index.md

---
title: "for...in"

> description: "Цикл for...in позволяет запустить цикл и пройти по ключам объекта"

Reviewer: Может так?
```suggestion
description: "Цикл for...in обходит перечисляемые ключи объекта"
```

authors:
  - windrushfarer
tags:
  - doka
---

## Кратко

> `for...in` позволяет в пройти в цикле по перечисляемым свойствам объекта, в том числе по свойствам из прототипа.

Reviewer: ```suggestion
`for...in` позволяет в пройти в [цикле](/js/loop) по перечисляемым свойствам объекта, в том числе по свойствам из прототипа.
```

> Перечисляемые свойства – это свойства, которые доступны для перебора в цикле `for...in`.

Reviewer: Тут закольцовка получилась, давай попробуем короткую эвристику придумать. Ниже есть хорошая фраза «все свойства, которые добавляется к объекту по умолчанию являются перечисляемыми», давай её перефразируем. 

Например, «Перечисляемые свойства — это свойства, которые программист добавляет объекту. Встроенные свойства (например, `length`) и методы не перечисляемые»

Author: Спасибо! Вышло неплохо

## Как пишется

Схематично структура для создания цикла выглядит так:

```js
for (переменная in объект) {
  // действия внутри цикла
}
```

> Для цикла необходимо объявить название переменной и указать сам объект, у которого будут перечисляться свойства. В объявленной переменной будет хранится название свойства во время итерации.

Reviewer: Мне кажется, что тут фраза «перечисляться свойства» может быть неправильно воспринято. Предлагаю более классический вариант:

 ```suggestion
Для цикла необходимо объявить название переменной и указать сам объект, свойства которого нужно обойти. В объявленной переменной будет храниться имя свойства во время итерации.
```

```js
const cat = {
  name: 'Борис',
  color: 'red',
  age: 8
}

for (const key in cat) {
  console.log(`${key} – ${cat[key]}`)
}
// Выведет в консоли:
// name – 'Борис',
// color – 'red',
// age – 8
```

## Как это понять

Цикл `for...in` хороший способ пройти по всем свойствам объекта, но стоит помнить несколько важных особенностей.

> ### Какие свойства будут перечислены

Reviewer: ```suggestion
### Что такое перечисляемые свойства
```

> В цикле будут только **перечисляемые** свойства объекта, это свойства, которые явно помечены, что они могут перечисляться в циклах. Например метод `indexOf()` у объекта `String` или метод `toString()` у [любого объекта](/js/objects-objects-everywhere) **не перечисляется** и потому они не участвуют в цикле `for...in`.

Reviewer: ```suggestion
В цикл попадут только _перечисляемые_ свойства объекта — это свойства, которые явно помечены, что они могут перечисляться в циклах. Например, метод [`indexOf()`](/js/index-of) у строки или метод [`toString()`](/js/object-tostring) у [любого объекта](/js/objects-objects-everywhere) не перечисляемые и не участвуют в цикле `for...in`.
```

Reviewer: Вообще тут нестыковка. Мы говорим «свойства« которые явно помечены перечисляемыми», но в следующем абзаце говори, что все свойства, которые добавляются объекту — перечисляемые. При этом мы нигде явно не пишем, что эти свойства перечиляемые

Author: Я не понял твою фразу, но мне абзац самому не очень нравится, потому перерфразировал

Базовое правило для запоминания – все свойства, которые добавляется к объекту по умолчанию являются перечисляемыми.

В цикле будут перечислены не только собственные свойства объекта, но и все перечисляемые свойства из прототипа объекта и прототипа прототипа и так далее.

```js
const a = { a: 1}
const b = { b: 2}
const c = { c: 3}

> b.__proto__ = a
> c.__proto__ = b

Reviewer: Использование `__proto__` уже deprecated, не стоит новичкам его показывать 
```suggestion
Object.setPrototypeOf(b, a)
Object.setPrototypeOf(c, b)
```

В цикле будут перечислены не только собственные свойства объекта, но и все перечисляемые свойства из прототипа объекта и прототипа прототипа и так далее.

```js
const a = { a: 1 }
const b = { b: 2 }
const c = { c: 3 }

b.__proto__ = a
c.__proto__ = b

for (const key in c) {
  console.log(key)
}
// Выведет: c, b, a
```

### Порядок перечисления

В стандарте языка нет описания как свойства должны перечисляться, потому это будет зависеть от реализации в конкретном движке браузера. Но есть базовые правила, которое можно встретить во всех браузерах:

1️⃣ Все строковые имена свойств будут перечисляться в порядке их присвоения к объекту:

```js
const developer = {
  name: 'Ваня',
  language: 'JavaScript',
  company: 'Google'
}

for (const key in developer) {
  console.log(key)
}
// Выведет: name, language, company
```

2️⃣ Все числовые свойства будут выводится в отсортированном порядке по возрастанию:

```js
const booksById = {
  341: {
    name: 'Harry Potter'
  },
  144: {
    name: 'Flowers for Algernon'
  },
  202: {
    name: 'Lord of the Rings'
  }
}

for (const key in booksById) {
  console.log(key)
}
// Выведет: 144, 202, 341
```

::: callout ❗️

> Несмотря на договорённости о порядке вывода числовых ключей, не нужно использовать `for...in` для перебора массива. В будущем текущие правила могут изменится. Для перебора у массива есть собственные методы: [`forEach`](/js/array-foreach), [`map`](/js/array-map) и другие. В этих методах порядок перебора гарантирован.

Reviewer 2: ```suggestion
Несмотря на договорённости о порядке вывода числовых ключей, не нужно использовать `for...in` для перебора массива. В будущем текущие правила могут измениться. Для перебора у массива есть собственные методы: [`forEach`](/js/array-foreach), [`map`](/js/array-map) и другие. В этих методах порядок перебора гарантирован.
```
